### ISSUE TITLE
strictRequired error when required properties exists and are declared inside oneOf

### ISSUE THREAD

**Issue by wdavidw** [ajv-validator/ajv#1571]
Using version 8.1.0 with `strict` mode enabled, an error of type `strictRequired` is thrown when `required` properties are declared inside `oneOf` even if those properties are defined next to `oneOf`.

**Ajv options object**

```javascript
const ajv = new Ajv({
  strict: true,
})
```

**JSON Schema**

```json
{
  "type": "object",
  "properties": {
    "property_a": {
      "type": "string",
    },
    "property_b": {
      "type": "string",
    },
  },
  "anyOf": [
    { "required": [ "property_a" ] },
    { "required": [ "property_b" ] }
  ]
}
```

**Sample data**

```json
{
  property_a: 'ok'
}
```

**Your code**

The code posted on [runkit](https://runkit.com/wdavidw/6087daa78dfb850013091d39) throw an error with the message `strict mode: required property "property_a" is not defined at "#/anyOf/0" (strictRequired)` (after switching from stack trace viewer to properties viewer in the UI)

**What results did you expect?**

Unless I am misunderstanding something, I expect the code to be valid since `property_a` is defined inside `properties`.

**Are you going to resolve the issue?**

This probably requires a deep understanding of Ajv internals.

**Comment by seriousme** [comment]
The code is valid, e.g.: 
```javascript

const ajv = new Ajv({
    strictRequired: false
});

const schema = {
	"type": "object",
	"properties": {
		"property_a": {
			"type": "string"
		},
		"property_b": {
			"type": "string"
		}
	},
	"anyOf": [{
			"required": ["property_a"]
		},
		{
			"required": ["property_b"]
		}
	]
};

const validate = ajv.compile(schema);
console.log(validate({property_a: "str"}))
console.log(validate({property_b: "str"}))
console.log(validate({property_c: "str"}))
```
Gives:
```javascript
true
true
false
```
as expected.

However the strict check gives out a warning. My guess is because `property_a` and `property_b` are officially only known within the properties branch and the `anyOf` branch is the start of a new subschema.

As the documentation mentions:
_By default Ajv executes in strict mode, that is designed to prevent any unexpected behaviours or silently ignored mistakes in schemas (see Strict Mode for more details). It does not change any validation results, but it makes some schemas invalid that would be otherwise valid according to JSON Schema specification._
https://ajv.js.org/options.html#strict-mode-options

So basically Ajv says: I think I understand what you mean with your schema, but if you want to be sure, adhere to the strict mode.

If you want to avoid the warning from 'strict mode' then restructuring the schema to:
```javascript
const ajv = new Ajv({ strict:true });
const schema = {
  "type": "object",
  "anyOf": [
    {
      "properties": {
        "property_a": {
          "type": "string"
        }
      },
      "required": ["property_a"] 
    },
    {
      "properties": {
        "property_b": {
          "type": "string"
        }
      },
      "required": ["property_b"] 
    }
  ]
};

const validate = ajv.compile(schema);
console.log(validate({ property_a: "str" }))
console.log(validate({ property_b: "str" }))
console.log(validate({ property_c: "str" }))
``` 
does the trick.

Hope this helps.

Kind regards,
Hans

**Comment by epoberezkin** [comment]
indeed. strictRequired is a bit limited to what it sees, it does top down analysis, but misses properties in siblings. It can be disabled without disabling the rest of the strict mode with `strictRequired: false`

**Comment by wdavidw** [comment]
Thank you both for your replies. Considering that this is the expected behavior, I am closing this issue.

**Comment by epoberezkin** [comment]
You're welcome. Let's keep it open - it's a limitation, not the optimal behaviour.

**Comment by LilyRose2798** [comment]
Is there a workaround for this that doesn't require `strictRequired: false` and that also works with Typescript? The solution by @seriousme fixes the schema compilation errors from ajv but causes type checking to fail due to not having the field be required in the top level of the schema. Is this something that can be resolved within the typings on this project by having it check that the field is required inside all sub-schema's in the `anyOf` rather than just on the root schema?

**Comment by serge-fabiano** [comment]
Have you planned a fix for this issue? When?

**Comment by michaelgwelch** [comment]
I think I'm seeing a similar issue when using `required` inside of an `if`.
